softlearning/environments/adapters/robosuite_adapter.py

```diff
@@ -0,0 +1,144 @@
+"""Implements a RobosuiteAdapter that converts Robosuite envs into SoftlearningEnv."""
+
+from collections import OrderedDict
+
+import numpy as np
+import robosuite as suite
+import gym
+from gym import spaces
+
+from .softlearning_env import SoftlearningEnv
+from softlearning.environments.adapters.gym_adapter import GymAdapter
+
+ROBOSUITE_ENVIRONMENTS = {}
+
+def convert_robosuite_to_gym_obs_space(robosuite_obs_space):
+    assert isinstance(robosuite_obs_space, OrderedDict)
+    listDict = []
+    for key, value in robosuite_obs_space.items():
+        listDict.append((key, spaces.Box(
+            low=-float("inf"),
+            high=float("inf"),
+            shape=(len(value),),
+            dtype=np.float32)))
+    return spaces.Dict(OrderedDict(listDict))
+
+def convert_robosuite_to_gym_action_space(robosuite_action_space):
+    assert isinstance(robosuite_action_space, tuple)
+    return spaces.Box(
+        low=robosuite_action_space[0],
+        high=robosuite_action_space[1],
+        dtype=np.float32)
+
+class RobosuiteAdapter(SoftlearningEnv):
+    """Adapter that implements the SoftlearningEnv for Robosuite envs."""
+
+    def __init__(self,
+                 domain,
+                 task,
+                 *args,
+                 env=None,
+                 normalize=True,
+                 observation_keys=None,
+                 unwrap_time_limit=True,
+                 **kwargs):
+        assert not args, (
+            "Robosuite environments don't support args. Use kwargs instead.")
+
+        self._Serializable__initialize(locals())
+
+        self.normalize = normalize
+        self.unwrap_time_limit = unwrap_time_limit
+
+        super(RobosuiteAdapter, self).__init__(domain, task, *args, **kwargs)
+
+        if env is None:
+            assert (domain is not None and task is not None), (domain, task)
+            env_id = f"{domain}{task}"
+            env = suite.make(env_id, **kwargs)
+        else:
+            assert domain is None and task is None, (domain, task)
+
+        # TODO(Alacarter): Check how robosuite handles max episode length termination.
+
+        assert isinstance(env.observation_spec(), OrderedDict), env.observation_spec()
+        self.observation_keys = observation_keys or tuple(env.observation_spec().keys())
+        assert set(self.observation_keys).issubset(set(env.observation_spec().keys())), (
+            (self.observation_keys, env.observation_spec().keys()))
+
+        if normalize:
+            np.testing.assert_equal(env.action_spec[0], -1, "Ensure spaces are normalized.")
```

Comment:
Could you add a test case for this?


```python
def test_fails_with_unnormalized_action_spec(self):
    env = EnvWithUnnormalizedActionSpec(...)
    with self.assertRaises(AssertionError):
        adapter = RobosuiteAdapter(env, ...)
```

Reply:
Hmm I'm not too sure how to get an environment that isn't normalized. I tried modifying the `action_spec` attribute of an existing instance of an environment to denormalize it, but I'm not able to set attribute values.

`>>> envs = ["SawyerLift", "SawyerStack", "SawyerPickPlace", "SawyerPickPlaceSingle", "SawyerPickPlaceMilk", "SawyerPickPlaceBread", "SawyerPickPlaceCereal", "SawyerPickPlaceCan", "SawyerNutAssembly", "SawyerNutAssemblySingle", "SawyerNutAssemblySquare", "SawyerNutAssemblyRound", "BaxterLift", "BaxterPegInHole"]
>>> for env_id in envs:
...     env = suite.make(env_id, has_renderer=False, has_offscreen_renderer=False, use_camera_obs=False)
...     print(env_id, env.action_spec)
... 
SawyerLift (array([-1., -1., -1., -1., -1., -1., -1., -1.]), array([1., 1., 1., 1., 1., 1., 1., 1.]))
SawyerStack (array([-1., -1., -1., -1., -1., -1., -1., -1.]), array([1., 1., 1., 1., 1., 1., 1., 1.]))
SawyerPickPlace (array([-1., -1., -1., -1., -1., -1., -1., -1.]), array([1., 1., 1., 1., 1., 1., 1., 1.]))
SawyerPickPlaceSingle (array([-1., -1., -1., -1., -1., -1., -1., -1.]), array([1., 1., 1., 1., 1., 1., 1., 1.]))
SawyerPickPlaceMilk (array([-1., -1., -1., -1., -1., -1., -1., -1.]), array([1., 1., 1., 1., 1., 1., 1., 1.]))
SawyerPickPlaceBread (array([-1., -1., -1., -1., -1., -1., -1., -1.]), array([1., 1., 1., 1., 1., 1., 1., 1.]))
SawyerPickPlaceCereal (array([-1., -1., -1., -1., -1., -1., -1., -1.]), array([1., 1., 1., 1., 1., 1., 1., 1.]))
SawyerPickPlaceCan (array([-1., -1., -1., -1., -1., -1., -1., -1.]), array([1., 1., 1., 1., 1., 1., 1., 1.]))
SawyerNutAssembly (array([-1., -1., -1., -1., -1., -1., -1., -1.]), array([1., 1., 1., 1., 1., 1., 1., 1.]))
SawyerNutAssemblySingle (array([-1., -1., -1., -1., -1., -1., -1., -1.]), array([1., 1., 1., 1., 1., 1., 1., 1.]))
SawyerNutAssemblySquare (array([-1., -1., -1., -1., -1., -1., -1., -1.]), array([1., 1., 1., 1., 1., 1., 1., 1.]))
SawyerNutAssemblyRound (array([-1., -1., -1., -1., -1., -1., -1., -1.]), array([1., 1., 1., 1., 1., 1., 1., 1.]))
BaxterLift (array([-1., -1., -1., -1., -1., -1., -1., -1., -1., -1., -1., -1., -1.,
       -1., -1., -1.]), array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.]))
BaxterPegInHole (array([-1., -1., -1., -1., -1., -1., -1., -1., -1., -1., -1., -1., -1.,
       -1.]), array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1., 1.]))
`

Reply:
The way you would test this is by creating your own *minimal* RobosuiteEnv. Something like:

```
from robosuite.environments import MujocoEnv

class MockEnv(MujocoEnv):
    # Fill in the methods that are required for minimal environment.
    @property
    def dof(self):
        return 5

    @property
    def action_spec(self):
        low, high = np.ones(self.dof) * -2.0, np.ones(self.dof) * 2.0
        return low, high
```

I'm not 100% sure what else is required for creating a minimal environment like this. Just try creating a test for an environment like this and see where it fails. Let me know if there's something you get stuck in.